tree/20_cfe_basics/time_lib.cf

```diff
@@ -0,0 +1,203 @@
+bundle agent date(test_timestamp)
+{
+  vars:
+    any::
+      "timestamp" int => now();
+    test::
+      "timestamp" int => "${test_timestamp}";
+
+  classes:
+    any::
+      "test" expression => regcmp("^\d+$", "${test_timestamp}");
+}
+
+# Store a unique host value based on IP + hostname
+# You can change its content if you want to run tests by calling this bundle with a parameter
+bundle agent splay(test_splay)
+{
+  vars:
+    any::
+      # maximum value of splay integer = 7 nibbles (2^28-1)
+      "splay_max" string => "268435455";
+      # pseudorandon deterministic hex number of 7 nibbles
+      "splay_hex" string => string_head(hash("${sys.host}${sys.ipv4}", "md5"), "7");
+      # decimal version of the splay integer
+      "splay_dec" string => execresult("/usr/bin/printf '%d' 0x${splay_hex}", "noshell");
+    test::
+      "splay_dec" string => "${test_splay}";
+
+  classes:
+    any::
+      "test" expression => regcmp("^\d+$", "${test_splay}");
+}
+
+# Generic Scheduler, manage periodicity and splay of execution
+#  The real start time is delayed by a constant number of minutes that is lower than the splay parameter.
+#  This splay time depends on the host properties (IP and hostname)
+#  Everything is calculated in minutes.
+#  This is the basis for other schedulers and is not made to be user friendly.
+#  The outcome are variables that can be used for final decision
+#
+# job_id: a string to identify this job
+# agent_periodicity: how often you run the agent in minutes
+# splay: on how many minutes you want to spread the runs
+# job_periodicity: the time frame between runs in minutes 
+# minutes_from_start: the number of minutes elapsed since the first start of the job (usually: now - any_start_epoch_timestamp/60)
+#
+# If there is an error, a class is raised: job_${job_id}_error
+#    # Sample code to use in the caller 
+#      "job_${job_id}_after_splay" expression => isgreaterthan("${job_schedule_generic.current_run_id}", "${job_schedule_generic.max_run_id}");
+#      "job_${job_id}_after_run" expression => isgreaterthan("${job_schedule_generic.current_run_id}", "${job_schedule_generic.job_run_id}");
+#      "job_${job_id}_run" expression => strcmp("${job_schedule_generic.current_run_id}", "${job_schedule_generic.job_run_id}");
+bundle agent schedule_generic(job_id, agent_periodicity, splay, job_periodicity, minutes_from_start)
+{
+  vars:
+    any::
+      "test" string => "value";
+    noerror::
+      # shift between each run = run_periodicity % agent_periodicity
+      "run_shift_real" string => eval("${job_periodicity} % ${agent_periodicity}", "math", "infix");
+      "run_shift" int => "${run_shift_real}";
+
+      # max_run_id = int(splay / agent periodicity) = min number of agent run in interval
+      "max_run_id_real" string => eval("${splay}/${agent_periodicity}", "math", "infix");
+      "max_run_id" int => "${max_run_id_real}";
```

Comment:
oh, it's explained. splay: on how many minutes you want to spread the runs

There are too many "splay" in this file. Could you give them more meaningful name ?